matrix-e2e.c

```diff
@@ -680,6 +680,119 @@ int matrix_e2e_get_device_keys(MatrixConnectionData *conn, const gchar *device_i
+    purple_debug_info("matrixprpl",
+                      "%s: Type %zd olm encrypted message from %s\n",
+                      __func__, (size_t)type, cevent_sender);
+    if (!type) {
+        /* A 'prekey' message to establish an Olm session
+         * TODO!!!!: Try existing sessions and check with
+         * matches_inbound_session */
+        session = olm_session(g_malloc0(olm_session_size()));
+        const gchar *cevent_body;
+        cevent_body = matrix_json_object_get_string_member(our_ciphertext,
+                                                               "body");
+        gchar *cevent_body_copy = g_strdup(cevent_body);
+        if (olm_create_inbound_session_from(session, conn->e2e->oa, sender_key,
+                                               strlen(sender_key),
+                                        cevent_body_copy, strlen(cevent_body))
+            == olm_error()) {
+            purple_debug_info("matrixprpl",
+                              "%s: prekey inbound_session_from failed : %s\n",
+                    __func__, olm_session_last_error(session));
+            goto err;
+        }
+        if (olm_remove_one_time_keys(conn->e2e->oa, session) == olm_error()) {
+            purple_debug_info("matrixprpl",
+                              "%s: Failed to remove 1tk, inbound session: %s\n",
+                    __func__, olm_account_last_error(conn->e2e->oa));
+            goto err;
+        }
+        cevent_body_copy = g_strdup(cevent_body);
+        size_t max_plaintext_len = olm_decrypt_max_plaintext_length(session,
+                                       0 /* Prekey */,
+                                       cevent_body_copy,
+                                       strlen(cevent_body_copy));
+        if (max_plaintext_len == olm_error()) {
+            purple_debug_info("matrixprpl",
+                              "%s: Failed to get plaintext length %s\n",
+                              __func__, olm_session_last_error(session));
+            goto err;
+        }
+        plaintext = g_malloc0(max_plaintext_len + 1);
+        cevent_body_copy = g_strdup(cevent_body);
+
+        size_t pt_len = olm_decrypt(session, 0 /* Prekey */, cevent_body_copy,
+                                       strlen(cevent_body),
+                                       plaintext, max_plaintext_len);
+        if (pt_len == olm_error() || pt_len >= max_plaintext_len) {
+            purple_debug_info("matrixprpl",
+                              "%s: Failed to decrypt inbound session creation"
+                              " event: %s\n",
+                              __func__, olm_session_last_error(session));
+            goto err;
+        }
+        plaintext[pt_len] = '\0';
+        // TODO: Store session in db
+    } else {
+        purple_debug_info("matrixprpl", "%s: Type %zd olm\n", __func__, type);
+    }
+    // TODO  resave account? Or just session?
+    //
+    g_free(plaintext);
```

Comment:
I've added a commit that adds:
/* Really clear an area of memory */
static void clear_mem(volatile char *data, size_t len)
{
#ifdef __STDC_LIB_EXT1__
    /* Untested! */
    memset_s(data, len, '\0', len);
#else
    size_t index;
    for(index = 0;index < len; index ++)
    {
        data[index] = '\0';
    }
#endif
}

and I can see at the end of the routine (inlined a few times):
    13b2: 48 89 e8              mov    %rbp,%rax
    13b5: 49 8d 14 2c           lea    (%r12,%rbp,1),%rdx
    13b9: 74 15                 je     13d0 <matrix_e2e_decrypt_d2d+0x350>
    13bb: c6 00 00              movb   $0x0,(%rax)
    13be: 48 83 c0 01           add    $0x1,%rax
    13c2: 48 39 d0              cmp    %rdx,%rax
    13c5: 75 f4                 jne    13bb <matrix_e2e_decrypt_d2d+0x33b>

that looks kind of believable - I'm not sure I completely trust it, still, seems better.